stream.go

```diff
@@ -140,6 +144,198 @@ func (r *Client) StreamPrediction(ctx context.Context, prediction *Prediction) (
 	return sseChan, errChan
 }
 
+func (r *Client) StreamPredictionFiles(ctx context.Context, prediction *Prediction) (<-chan streaming.File, error) {
+	url := prediction.URLs["stream"]
+	if url == "" {
+		return nil, errors.New("streaming not supported or not enabled for this prediction")
+	}
+
+	ch := make(chan streaming.File)
+
+	go r.streamFilesTo(ctx, ch, url, "")
+	return ch, nil
+}
+
+func (r *Client) StreamPredictionText(ctx context.Context, prediction *Prediction) (io.Reader, error) {
+	url := prediction.URLs["stream"]
+	if url == "" {
+		return nil, errors.New("streaming not supported or not enabled for this prediction")
+	}
+
+	reader, writer := io.Pipe()
+
+	go r.streamTextTo(ctx, writer, url, "")
+	return reader, nil
+}
+
+func (r *Client) streamTextTo(ctx context.Context, writer *io.PipeWriter, url string, lastEventID string) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		writer.CloseWithError(fmt.Errorf("failed to create request: %w", err))
+		return
+	}
+	req.Header.Set("Accept", "text/event-stream")
+	req.Header.Set("Cache-Control", "no-cache")
+	req.Header.Set("Connection", "keep-alive")
+
+	if lastEventID != "" {
+		req.Header.Set("Last-Event-ID", lastEventID)
+	}
+
+	resp, err := r.c.Do(req)
+	if err != nil {
+		writer.CloseWithError(fmt.Errorf("failed to send request: %w", err))
+		return
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		writer.CloseWithError(fmt.Errorf("received invalid status code: %d", resp.StatusCode))
+		return
+	}
+	defer resp.Body.Close()
+	decoder := eventsource.NewDecoder(resp.Body)
+	for {
+		event, err := decoder.Decode()
+		if err != nil {
+			if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+				// retry (TODO: backoff policy?)
+				r.streamTextTo(ctx, writer, url, lastEventID)
+				return
+			}
+			writer.CloseWithError(fmt.Errorf("Failed to get token: %w", err))
+			return
+		}
+		lastEventID = event.Id()
+		switch event.Event() {
+		case SSETypeOutput:
+			io.WriteString(writer, event.Data())
+		case SSETypeDone:
+			writer.Close()
```

Comment:
Do we need to do any other teardown here? Of the client request?

Reply:
We have a `defer resp.Body.Close()` on line 193, no?